### PR TITLE
Add data persistence and refactor tests

### LINKED ISSUE #10: The user should be able to see messages sent while he was offline
Estimate: 1

### LINKED ISSUE #11: Users should be authenticated
Estimate: 1

### PR DESCRIPTION
**Pull request by piecubed**

### CODE REVIEW

**Comment by piecubed** [comment]
Weird debugging thing was because of an issue only on GitHub actions, not on my computer.

**Comment by JosiahWI** [comment]
There still seems to be a timing issue on the windows-latest CI. We just removed Windows and MacOSX from the CI to reduce CI minutes, but maybe we should consider adding a Windows build back in on both the server and client.

File: src/server/handlers.go [hunk 1]
```diff
@@ -2,61 +2,104 @@ package server
...
+		return
+	}
+
+	msg_broadcast := websocket_models.BroadcastMessage{
+		Time: time.Now().Unix(),
+		Author: websocket_models.Author{
+			ID:       client.User.UserID.Hex(),
+			Username: client.User.Username,
+		},
 		Message: msg.Message,
```
Comment (JosiahWI) [inline src/server/handlers.go:42]: Message content needs to be tested.

File: src/server/handlers.go [hunk 2]
```diff
@@ -2,61 +2,104 @@ package server
...
-	m.hub.HubChannels.broadcast <- recv_msg
+	db_msg := database.Message{
+		Content:   msg_broadcast.Message,
+		Timestamp: msg_broadcast.Time,
+		Author:    msg_broadcast.Author.ID,
+	}
+
+	err = m.hub.Database.InsertMessage(&db_msg)
+	if err != nil {
+		client.OutgoingPayloadQueue <- websocket_models.GenericError{Error: "DatabaseError"}
```
Comment (JosiahWI) [inline src/server/handlers.go:54]: Test this.

File: src/server/handlers.go [hunk 3]
```diff
@@ -2,61 +2,104 @@ package server
...
+	}
+
+	msg_broadcast.MessageID = db_msg.MessageID.Hex()
```
Comment (JosiahWI) [inline src/server/handlers.go:57]: Test this.

File: src/server/handlers.go [hunk 4]
```diff
@@ -2,61 +2,104 @@ package server
...
+	m.hub.broadcast_payload <- msg_broadcast
 }
 
+func (m *MessageHandler) HandleMessageHistory(b []byte, client *Client) {
+	hist := &websocket_models.MessageHistory{}
+	json.Unmarshal(b, hist)
+	if hist.From == 0 || hist.To == 0 {
```
Comment (JosiahWI) [inline src/server/handlers.go]: Test this.

File: src/server/server_hub.go [hunk 5]
```diff
@@ -202,8 +209,14 @@ func Serve(addr string, wg *utils.WaitGroupCounter, hub *ServerHub) {
...
-func HandleFunc(hub *ServerHub, wg *utils.WaitGroupCounter) http.HandlerFunc {
+func (hub *ServerHub) HTTPRequestHandler() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		hub.Upgrade(w, r, wg)
+		if r.URL.Path == "/login" {
+			hub.Login(w, r)
+		} else if r.URL.Path == "/register" {
+			hub.Register(w, r)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
```
Comment (JosiahWI) [inline src/server/server_hub.go:217]: Test this.

**Review by JosiahWI — CHANGES_REQUESTED** [review]
Besides all the mentioned tests, please remember to run `go fmt`.

**Comment by piecubed** [comment]
I did run gofmt.

File: src/server/handlers.go (hunk 1 above)
Comment (piecubed) [reply src/server/handlers.go:42]: It is.

**Review by JosiahWI — CHANGES_REQUESTED** [review]
Please add a test that the message IDs are valid.

**Review by JosiahWI — APPROVED** [review]

**Pull request merged**